accessible/tree.service.js

```diff
@@ -0,0 +1,325 @@
+      this.activeDesc_[id] = node;
+    },
+    getActiveDesc: function(id) {
+      return this.activeDesc_[id];
+    },
+    updateSelectedNode: function(node, tree, keepFocus) {
+      blocklyApp.debug && console.log('updating node: ' + node.id);
+      var treeId = tree.id;
+      var activeDesc = this.getActiveDesc(treeId);
+      if (activeDesc) {
+        activeDesc.classList.remove('blocklyActiveDescendant');
+      } else {
+        blocklyApp.debug && console.log('updateSelectedNode: there is no active descendant');
+      }
+      node.classList.add('blocklyActiveDescendant');
+      tree.setAttribute('aria-activedescendant', node.id);
+      this.setActiveDesc(node, treeId);
+      node.setAttribute('aria-selected', 'true');
+
+      // make sure keyboard focus is on tree as a whole
+      // in case before the user was editing a block and keyboard
+      // focus got shifted.
+      if (keepFocus) {
+        tree.focus();
+      }
+    },
+    workspaceButtonKeyHandler: function(e, treeId) {
+      blocklyApp.debug && console.log(e.keyCode);
+      blocklyApp.debug && console.log('inside TreeService');
+      switch (e.keyCode) {
+        case 9:
+          // 16,9: shift, tab
+          if (e.shiftKey) {
+            console.log('shifttabbing');
+            // if the previous key is shift, we're shift-tabbing mode
+            this.goToPreviousTree(treeId, e);
+          } else {
+            // if previous key isn't shift, we're tabbing
+            this.goToNextTree(treeId, e);
+          }
+          break;
+      }
+      this.previousKey_ = e.keyCode;
+    },
+    goToNextTree: function(treeId, e) {
+      var next = false;
+      for (var i = 0; i < this.trees.length; i++) {
+        if (next) {
+          this.trees[i].focus();
+          e.preventDefault();
+          e.stopPropagation();
+          return;
+        }
+        if (this.trees[i].id == treeId) {
+          next = true;
+        }
+      }
+      e.preventDefault();
+      e.stopPropagation();
+    },
+    goToPreviousTree: function(treeId, e) {
+      if (treeId == this.trees[0].id) {
+        return;
+      }
+      // otherwise, go to the previous tree class
+      var next = false;
+      for (var i = (this.trees.length - 1); i >= 0; i--) {
+        if (next) {
+          this.trees[i].focus();
+          e.preventDefault();
+          e.stopPropagation();
+          return;
+        }
+        if (this.trees[i].id == treeId) {
+          next = true;
+        }
+      }
+      e.preventDefault();
+      e.stopPropagation();
+    },
+    keyHandler: function(e, tree) {
+      var treeId = tree.id;
+      var node = this.getActiveDesc(treeId);
+      if (!node) {
+        blocklyApp.debug && console.log('KeyHandler: no active descendant');
+      }
+      blocklyApp.debug && console.log(e.keyCode);
+      blocklyApp.debug && console.log('inside TreeService');
+      switch (e.keyCode) {
+        case 9:
+          // 16,9: shift, tab
```

Comment:
I was just noting what the key codes for shift and tab are so that others maintaining this codebase know what keycodes to look out for in debug mode.

Will look into fixing duplication of code.